cmd/restic/cmd_copy.go

```diff
@@ -0,0 +1,203 @@
+	dstGopts.password, err = resolvePassword(dstGopts, "RESTIC_PASSWORD2")
+	if err != nil {
+		return err
+	}
+	dstGopts.password, err = ReadPassword(dstGopts, "enter password for destination repository: ")
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithCancel(gopts.ctx)
+	defer cancel()
+
+	srcRepo, err := OpenRepository(gopts)
+	if err != nil {
+		return err
+	}
+
+	dstRepo, err := OpenRepository(dstGopts)
+	if err != nil {
+		return err
+	}
+
+	srcLock, err := lockRepo(srcRepo)
+	defer unlockRepo(srcLock)
+	if err != nil {
+		return err
+	}
+
+	dstLock, err := lockRepo(dstRepo)
+	defer unlockRepo(dstLock)
+	if err != nil {
+		return err
+	}
+
+	debug.Log("Loading source index")
+	if err := srcRepo.LoadIndex(ctx); err != nil {
+		return err
+	}
+
+	debug.Log("Loading destination index")
+	if err := dstRepo.LoadIndex(ctx); err != nil {
+		return err
+	}
+
+	for sn := range FindFilteredSnapshots(ctx, srcRepo, opts.Hosts, opts.Tags, opts.Paths, args) {
+		Verbosef("snapshot %s of %v at %s)\n", sn.ID().Str(), sn.Paths, sn.Time)
+		Verbosef("  copy started, this may take a while...\n")
+
+		if err := copySnapshot(ctx, srcRepo, dstRepo, *sn.Tree); err != nil {
+			return err
+		}
+		debug.Log("tree copied")
+
+		if err = dstRepo.Flush(ctx); err != nil {
+			return err
+		}
+		debug.Log("flushed packs")
+
+		err = dstRepo.SaveIndex(ctx)
+		if err != nil {
+			debug.Log("error saving index: %v", err)
+			return err
+		}
+		debug.Log("saved index")
+
+		// save snapshot
+		sn.Parent = nil   // Parent does not have relevance in the new repo.
+		sn.Original = nil // Original does not have relevance in the new repo.
+		newID, err := dstRepo.SaveJSONUnpacked(ctx, restic.SnapshotFile, sn)
+		if err != nil {
+			return err
+		}
+		Verbosef("snapshot %s saved\n", newID.Str())
+	}
+	return nil
+}
+
+func copySnapshot(ctx context.Context, srcRepo, dstRepo restic.Repository, treeID restic.ID) error {
+	tree, err := srcRepo.LoadTree(ctx, treeID)
+	if err != nil {
+		return fmt.Errorf("LoadTree(%v) returned error %v", treeID.Str(), err)
+	}
+	// Do we already have this tree blob?
+	if !dstRepo.Index().Has(treeID, restic.TreeBlob) {
+		newTreeID, err := dstRepo.SaveTree(ctx, tree)
+		if err != nil {
+			return fmt.Errorf("SaveTree(%v) returned error %v", treeID.Str(), err)
+		}
+		// Assurance only.
+		if newTreeID != treeID {
+			return fmt.Errorf("SaveTree(%v) returned unexpected id %s", treeID.Str(), newTreeID.Str())
+		}
+	}
+
+	// TODO: keep only one (big) buffer around.
+	// TODO: parellize this stuff, likely only needed inside a tree.
+
+	for _, entry := range tree.Nodes {
+		// If it is a directory, recurse
+		if entry.Type == "dir" && entry.Subtree != nil {
+			if err := copySnapshot(ctx, srcRepo, dstRepo, *entry.Subtree); err != nil {
+				return err
+			}
+		}
+		// Copy the blobs for this file.
+		for _, blobID := range entry.Content {
+			// Do we already have this data blob?
+			if dstRepo.Index().Has(blobID, restic.DataBlob) {
```

Comment:
This currently doesn't work as expected: Newly uploaded blobs don't show up in the repository index until their pack is uploaded. This can lead to duplicate blobs created by the copy command. The archiver code currently uses a sort of workaround to manually track these new blobs. It would be possible (and rather easy) to add similar code here. But the best solution would be move this functionality into the MasterIndex similar to the `CheckSetKnown` method from the index rewrite issue https://github.com/restic/restic/issues/2523#issuecomment-585940174 https://github.com/aawsome/restic/commit/53f8a9980cb2e52beb453677304b565e075427f5#diff-0a50833756dabb93f6aaa532c9f63ec6

Reply:
As I don't see that #2523 will be tackled soon (and so far there are quite a lot open things to work on there), maybe it is best to make a small PR only implementing the 'CheckSetKnown' part of #2523. 

I can propose to prepare such an PR which would then be a prerequisite for this PR.

Reply:
To be precise for the current implementation in this PR we wouldn't need an atomic test&set method for the index. It would suffice if `SaveBlob` remembered which blocks it has already saved. If we want to parallelize the implementation then we need something like the `CheckSetKnown` method.

However, there's another option: Let `SaveBlob` itself check for duplicates. This would conveniently hide the details on how to properly check for blobs that are in incomplete packs and by construction ensure that we can't accidentally call `CheckSetKnown` and then fail to actually save the chunk.

Reply:
This should be now resolved with #2773.